### PR TITLE
Implement ursacli, a native client

### PR DESCRIPTION
**Pull request by icedevml**
* Native ursadb client called `ursacli`
* Interactive mode / single command mode (supplied from argv)
* Progress reporting (additional thread which calls `status;` every second)

image

### CODE REVIEW

File: src/Client.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        auto res_tasks = res["result"]["tasks"];
+
+        for (json::iterator it = res_tasks.begin(); it != res_tasks.end();
+             ++it) {
+            if ((*it)["connection_id"] == conn_id) {
+                unsigned int work_done = (*it)["work_done"].get<unsigned int>();
+                unsigned int work_estimated =
+                    (*it)["work_estimated"].get<unsigned int>();
+                unsigned int work_perc =
+                    work_estimated > 0 ? work_done * 100 / work_estimated : 0;
```
Comment (chivay) [inline src/Client.cpp:42]: Please, use sized integers (`uint32_t`/`uint64_t`) or `int`

File: src/Client.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    while (true) {
+        zmq::context_t context(1);
+        zmq::socket_t socket(context, ZMQ_REQ);
+        socket.setsockopt(ZMQ_LINGER, 0);
+        socket.setsockopt(ZMQ_RCVTIMEO, 1000);
+        socket.connect(server_addr);
+
+        s_send(socket, "status;");
+        std::string res_str;
+        unsigned int retries = 0;
```
Comment (chivay) [inline src/Client.cpp:19]: same as below

File: .github/workflows/test_code.yml [hunk 3]
```diff
@@ -21,7 +21,7 @@ jobs:
     - name: clang-format
       run: cmake --build build --target format
     - name: check formatting
-      run: git diff --quiet
+      run: git diff --quiet || ( git diff && false )
```
Comment (chivay) [inline .github/workflows/test_code.yml:24]: Not sure if we want this. The output of the command isn't relevant at all. `make format` takes care of fixing it.
But if we do, one call to `git diff --exit-code` would be enough

File: src/Client.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+int main(int argc, char *argv[]) {
+    std::string server_addr = "tcp://localhost:9281";
+    std::string db_command = "";
+
+    if (argc >= 2) {
+        server_addr = std::string(argv[1]);
+
+        if (server_addr == "-h" || server_addr == "--help") {
```
Comment (chivay) [inline src/Client.cpp:59]: This looks really weird. I would move the assignment outside of if statement

File: src/Client.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            auto res_type = res["type"].get<std::string>();
+
+            if (res_type == "error") {
+                spdlog::error(res["error"]["message"].get<std::string>());
+            } else {
+                spdlog::info(res.dump(4));
+            }
+
+            break;
+        } while (1);
```
Comment (chivay) [inline src/Client.cpp:141]: Maybe move this section to a separate function?

File: .github/workflows/test_code.yml (hunk 3 above)
Comment (icedevml) [reply .github/workflows/test_code.yml:24]: It's always better to see what are the exact "complaints" of the formatter. In my case it was just a single line to fix, so it was even easier to edit by hand than running `make format` locally.

Comment (chivay) [reply .github/workflows/test_code.yml:24]: OK, so let's use `git diff --exit-code` then

Comment (icedevml) [reply .github/workflows/test_code.yml:24]: How about that?

```
    - name: check formatting
      run: git diff --exit-code || ( >&2 echo "Please run 'make format' to fix these issues automatically." && false )
```

Comment (chivay) [reply .github/workflows/test_code.yml:24]: Sure :>

File: src/Client.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            }
+
+            res_str = s_recv(socket);
+            retries++;
+        }
+
+        auto res = json::parse(res_str);
+        auto res_tasks = res["result"]["tasks"];
+
+        for (json::iterator it = res_tasks.begin(); it != res_tasks.end();
```
Comment (msm-code) [inline src/Client.cpp:35]: With our json lib it's usually not needed to use raw iterators. Try: 

```cpp
for (const auto &it: res_tasks) {
} 
```

(not sure if it'll work literally like this, but in principle something like this should).

File: src/Client.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+             ++it) {
+            if ((*it)["connection_id"] == conn_id) {
```
Comment (msm-code) [inline src/Client.cpp:37]: Maybe `if ((*it)["connection_id"] != conn_id) { continue; } ` to reduce nesting?

Optional though, whichever you prefer.

File: src/Client.cpp [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                unsigned int work_done = (*it)["work_done"].get<unsigned int>();
```
Comment (msm-code) [inline src/Client.cpp:38]: Is the .get() necessary? Json is usually smart, so

```cpp
unsigned int work_done = (*it)["work_done"];
```

should work as well. The same applies below.

File: src/Client.cpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                "if not provided - interactive mode");
+            return 0;
+        }
+    }
+
+    if (argc >= 3) {
+        db_command = std::string(argv[2]);
+    }
+
+    if (argc >= 4) {
```
Comment (msm-code) [inline src/Client.cpp:75]: Print usage maybe? (If yes, than probably move usage() to a helper function).

File: src/Client.cpp [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    auto res_str = s_recv(socket);
+
+    if (res_str.empty()) {
+        spdlog::error("Failed to connect to the database!");
+        return 1;
+    }
+
+    auto res = json::parse(res_str)["result"];
+
+    std::string server_version = res["ursadb_version"].get<std::string>();
```
Comment (msm-code) [inline src/Client.cpp:96]: I'm pretty sure get() here is unnecessary.

File: src/Client.cpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+            // execute single command and exit
+            s_send(socket, db_command);
+        } else {
+            // interactive mode
+            std::cout << "ursadb> ";
+
+            std::string cmd;
+            std::getline(std::cin, cmd);
+
+            s_send(socket, cmd);
```
Comment (msm-code) [inline src/Client.cpp:121]: tricky part - in concern of user friendliness, append `;` to the cmd (do I remember correctly, that just `ping` without `;` will cause a parse error?

File: src/Client.cpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+        do {
+            auto res_str = s_recv(socket);
+
+            if (res_str.empty()) {
+                continue;
+            }
+
+            auto res = json::parse(res_str);
+            auto res_type = res["type"].get<std::string>();
```
Comment (msm-code) [inline src/Client.cpp:132]: std::string res_type and get is unnecessary.

**Pull request merged**